tests/robustness/multiclient_test/framework/snapshotter.go

```diff
@@ -0,0 +1,192 @@
+}
+
+// RestoreSnapshot delegates to a specific client's Snapshotter.
+func (mcs *MultiClientSnapshotter) RestoreSnapshot(ctx context.Context, snapID, restoreDir string, opts map[string]string) (fingerprint []byte, err error) {
+	ks, err := mcs.createOrGetSnapshotter(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return ks.RestoreSnapshot(ctx, snapID, restoreDir, opts)
+}
+
+// RestoreSnapshotCompare delegates to a specific client's Snapshotter.
+func (mcs *MultiClientSnapshotter) RestoreSnapshotCompare(ctx context.Context, snapID, restoreDir string, validationData []byte, reportOut io.Writer, opts map[string]string) (err error) {
+	ks, err := mcs.createOrGetSnapshotter(ctx)
+	if err != nil {
+		return err
+	}
+
+	return ks.RestoreSnapshotCompare(ctx, snapID, restoreDir, validationData, reportOut, opts)
+}
+
+// DeleteSnapshot delegates to a specific client's Snapshotter.
+func (mcs *MultiClientSnapshotter) DeleteSnapshot(ctx context.Context, snapID string, opts map[string]string) error {
+	ks, err := mcs.createOrGetSnapshotter(ctx)
+	if err != nil {
+		return err
+	}
+
+	return ks.DeleteSnapshot(ctx, snapID, opts)
+}
+
+// RunGC runs garbage collection on the server repository directly since clients
+// are not authorized to do this.
+func (mcs *MultiClientSnapshotter) RunGC(ctx context.Context, opts map[string]string) error {
+	return mcs.server.RunGC(ctx, opts)
+}
+
+// ListSnapshots delegates to a specific client's Snapshotter.
+func (mcs *MultiClientSnapshotter) ListSnapshots(ctx context.Context) ([]string, error) {
+	ks, err := mcs.createOrGetSnapshotter(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return ks.ListSnapshots(ctx)
+}
+
+// Cleanup delegates to a specific client's Snapshotter Cleanup method, but also
+// disconnects the client from the server, removes the client from the server's
+// user list, and removes the ClientSnapshotter from MultiClientSnapshotter.
+func (mcs *MultiClientSnapshotter) Cleanup() {
+	for clientID, s := range mcs.clients {
+		s.DisconnectClient(clientID)
+		s.Cleanup()
+		mcs.server.RemoveClient(clientID)
+
+		delete(mcs.clients, clientID)
+	}
+
+	mcs.server.Cleanup()
+}
+
+// createOrGetSnapshotter gets a client's Snapshotter from the given context if
+// possible or creates a new ClientSnapshotter.
+func (mcs *MultiClientSnapshotter) createOrGetSnapshotter(ctx context.Context) (robustness.Snapshotter, error) {
+	c := UnwrapContext(ctx)
+	if c == nil {
+		log.Println("Context does not contain a client")
+		return nil, robustness.ErrKeyNotFound
+	}
+
+	log.Printf("Issuing Kopia command for client: %s\n", c.ID)
+
+	// Get existing ClientSnapshotter if available
+	mcs.mu.RLock()
+	cs, ok := mcs.clients[c.ID]
+	mcs.mu.RUnlock()
+
+	if ok {
+		return cs, nil
+	}
+
+	// Create new ClientSnapshotter
+	mcs.mu.Lock()
+	defer mcs.mu.Unlock()
+
+	// Check if ClientSnapshotter was created while lock was open
+	cs, ok = mcs.clients[c.ID]
+	if ok {
+		return cs, nil
+	}
```

Comment:
I don't think this assumption is realistic for this set up: there is only one call at a time to the harness for a given client. 

I suggest not acquiring the write lock until you've constructed the client and are ready to assign it to the `mcs.clients` map - then you may do your paranoia check if you wish.